src/EditorFeatures/Core.Wpf/InheritanceMargin/InheritanceChainMarginTaggerProvider.cs

```diff
@@ -0,0 +1,104 @@
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.InheritanceMargin;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InheritanceMargin
+{
+    [Export(typeof(IViewTaggerProvider))]
+    [TagType(typeof(InheritanceMarginTag))]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [ContentType(ContentTypeNames.VisualBasicContentType)]
+    [Name(nameof(InheritanceChainMarginTaggerProvider))]
+    internal sealed class InheritanceChainMarginTaggerProvider : AsynchronousViewTaggerProvider<InheritanceMarginTag>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InheritanceChainMarginTaggerProvider(
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            IForegroundNotificationService notificationService) : base(
+                threadingContext,
+                listenerProvider.GetListener(FeatureAttribute.InheritanceChainMargin),
+                notificationService)
+        {
+        }
+
+        protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnWorkspaceChanged(subjectBuffer, TaggerDelay.OnIdle, AsyncListener),
+                TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.OnIdle),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, InheritanceMarginOptions.ShowInheritanceMargin, TaggerDelay.OnIdle));
+        }
+
+        protected override async Task ProduceTagsAsync(
+            TaggerContext<InheritanceMarginTag> context,
+            DocumentSnapshotSpan spanToTag,
+            int? caretPosition)
+        {
+            var document = spanToTag.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+            var inheritanceMarginInfoService = document.GetLanguageService<IInheritanceMarginService>();
+            if (inheritanceMarginInfoService == null)
+            {
+                return;
+            }
+
+            var inheritanceMemberItems = await inheritanceMarginInfoService
+                .GetInheritanceInfoAsync(
+                    document,
+                    cancellationToken).ConfigureAwait(false);
```

Comment:
really wonky formatting.  in general, this would look like:

```suggestion
            var inheritanceMemberItems = await inheritanceMarginInfoService.GetInheritanceInfoAsync(
                    document, cancellationToken).ConfigureAwait(false);
```